mithril-client-wasm/src/certificate_verification_cache.rs

```diff
@@ -0,0 +1,441 @@
+use anyhow::{anyhow, Context};
+use async_trait::async_trait;
+use chrono::{DateTime, TimeDelta, Utc};
+use std::ops::Add;
+use web_sys::{window, Storage};
+
+use mithril_client::certificate_client::CertificateVerifierCache;
+use mithril_client::MithrilResult;
+
+pub type CertificateHash = str;
+pub type PreviousCertificateHash = str;
+
+/// Browser local-storage based cache for the certificate verifier.
+///
+/// Note : as this cache is based on the browser local storage, it can only be used in a browser
+/// (it is not compatible with nodejs or other non-browser environment).
+pub struct LocalStorageCertificateVerifierCache {
+    cache_key_prefix: String,
+    expiration_delay: TimeDelta,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, serde::Serialize, serde::Deserialize)]
+struct CachedCertificate {
+    previous_hash: String,
+    expire_at: DateTime<Utc>,
+}
+
+impl CachedCertificate {
+    fn new<TPreviousHash: Into<String>>(
+        previous_hash: TPreviousHash,
+        expire_at: DateTime<Utc>,
+    ) -> Self {
+        CachedCertificate {
+            previous_hash: previous_hash.into(),
+            expire_at,
+        }
+    }
+}
+
+impl LocalStorageCertificateVerifierCache {
+    /// `LocalStorageCertificateVerifierCache` factory
+    pub fn new(cache_key_prefix_seed: &str, expiration_delay: TimeDelta) -> Self {
+        const CACHE_KEY_BASE_PREFIX: &'static str = "certificate_cache";
+
+        LocalStorageCertificateVerifierCache {
+            cache_key_prefix: format!("{CACHE_KEY_BASE_PREFIX}_{cache_key_prefix_seed}_"),
+            expiration_delay,
+        }
+    }
+
+    fn push(
+        &self,
+        certificate_hash: &CertificateHash,
+        previous_certificate_hash: &PreviousCertificateHash,
+        expire_at: DateTime<Utc>,
+    ) -> MithrilResult<()> {
+        if let Some(storage) = open_local_storage()? {
+            let key = self.cache_key(certificate_hash);
+            storage
+                .set_item(
+                    &key,
+                    &serde_json::to_string(&CachedCertificate::new(
+                        previous_certificate_hash,
+                        expire_at,
+                    ))
+                    .map_err(|err| anyhow!("Error serializing cache: {err:?}"))?,
+                )
+                .map_err(|err| anyhow!("Error storing key `{key}` in local storage: {err:?}"))?;
+        }
+        Ok(())
+    }
+
+    fn parse_cached_certificate(value: String) -> MithrilResult<CachedCertificate> {
+        serde_json::from_str(&value)
+            .map_err(|err| anyhow!("Error deserializing cached certificate: {err:?}"))
+    }
+
+    fn cache_key(&self, certificate_hash: &CertificateHash) -> String {
+        format!("{}{}", self.cache_key_prefix, certificate_hash)
+    }
+}
+
+fn open_local_storage() -> MithrilResult<Option<Storage>> {
+    let window = window()
```

Comment:
I changed this api to not return an option but instead raise an error if the local storage is missing. This should not be problematic since when building this cache we verify beforehand that the local storage is available and don't build the cache plus log an warning if it's not.